### PR TITLE
feat(bitmex): watchPositions && parsePosition fix

### PR DESCRIPTION
**Pull request by pcriadoperez**

### CODE REVIEW

**Comment by pcriadoperez** [comment]
###python
```python
[{'collateral': None,
  'contractSize': 100000000.0,
  'contracts': 400.0,
  'datetime': '2023-12-07T00:24:00.775Z',
  'entryPrice': None,
  'hedged': None,
  'id': '412475',
  'info': {'account': 412475,
           'currency': 'XBt',
           'currentQty': 400,
           'homeNotional': 0.00912904,
           'liquidationPrice': 20976.0,
           'markPrice': 43816.25,
           'markValue': -912904,
           'posCross': 635,
           'posMargin': 10450,
           'riskValue': 912904,
           'symbol': 'XBTUSD',
           'timestamp': '2023-12-07T00:24:00.775Z',
           'unrealisedPnl': -635,
           'unrealisedRoePcnt': -0.0696},
  'initialMargin': None,
  'initialMarginPercentage': None,
  'lastPrice': None,
  'lastUpdateTimestamp': None,
  'leverage': None,
  'liquidationPrice': 20976.0,
  'maintenanceMargin': None,
  'maintenanceMarginPercentage': None,
  'marginMode': 'isolated',
  'marginRatio': None,
  'markPrice': 43816.25,
  'notional': 0.00912904,
  'percentage': None,
  'side': 'long',
  'stopLossPrice': None,
  'symbol': 'BTC/USD:BTC',
  'takeProfitPrice': None,
  'timestamp': 1701908640775,
  'unrealizedPnl': '-0.00000635'}]

```
###php
```php
Array
(
    [0] => Array
        (
            [info] => Array
                (
                    [account] => 412475
                    [symbol] => XBTUSD
                    [currency] => XBt
                    [underlying] => XBT
                    [quoteCurrency] => USD
                    [commission] => 0.00075
                    [initMarginReq] => 0.01
                    [maintMarginReq] => 0.0035
                    [riskLimit] => 20000000000
                    [leverage] => 100
                    [crossMargin] => 1
                    [deleveragePercentile] => 1
                    [rebalancedPnl] => 0
                    [prevRealisedPnl] => 0
                    [prevUnrealisedPnl] => 0
                    [openingQty] => 400
                    [openOrderBuyQty] => 0
                    [openOrderBuyCost] => 0
                    [openOrderBuyPremium] => 0
                    [openOrderSellQty] => 0
                    [openOrderSellCost] => 0
                    [openOrderSellPremium] => 0
                    [currentQty] => 400
                    [currentCost] => -912269
                    [currentComm] => 684
                    [realisedCost] => 0
                    [unrealisedCost] => -912269
                    [grossOpenPremium] => 0
                    [isOpen] => 1
                    [markPrice] => 43824.4
                    [markValue] => -912732
                    [riskValue] => 912732
                    [homeNotional] => 0.00912732
                    [foreignNotional] => -400
                    [posCost] => -912269
                    [posCross] => 463
                    [posComm] => 692
                    [posLoss] => 0
                    [posMargin] => 10278
                    [posMaint] => 3885
                    [initMargin] => 0
                    [maintMargin] => 9815
                    [realisedPnl] => -684
                    [unrealisedPnl] => -463
                    [unrealisedPnlPcnt] => -0.0005
                    [unrealisedRoePcnt] => -0.0508
                    [avgCostPrice] => 43846.7643
                    [avgEntryPrice] => 43846.7643
                    [breakEvenPrice] => 43880
                    [marginCallPrice] => 20976
                    [liquidationPrice] => 20976
                    [bankruptPrice] => 20941
                    [timestamp] => 2023-12-07T00:24:30.768Z
                )

            [id] => 412475
            [symbol] => BTC/USD:BTC
            [timestamp] => 1701908670768
            [datetime] => 2023-12-07T00:24:30.768Z
            [lastUpdateTimestamp] => 
            [hedged] => 
            [side] => long
            [contracts] => 400
            [contractSize] => 100000000
            [entryPrice] => 43846.7643
            [markPrice] => 43824.4
            [lastPrice] => 
            [notional] => 400
            [leverage] => 100
            [collateral] => 
            [initialMargin] => 0
            [initialMarginPercentage] => 0.01
            [maintenanceMargin] => 0.00009815
            [maintenanceMarginPercentage] => 0.0035
            [unrealizedPnl] => -0.00000463
            [liquidationPrice] => 20976
            [marginMode] => cross
            [marginRatio] => 
            [percentage] => -0.0005
            [stopLossPrice] => 
            [takeProfitPrice] => 
        )
)
```

File: ts/src/pro/bitmex.ts [hunk 1]
```diff
@@ -603,6 +604,212 @@ export default class bitmex extends bitmexRest {
         }
     }
 
+    async watchPositions (symbols: Strings = undefined, since: Int = undefined, limit: Int = undefined, params = {}) {
```
Comment (carlosmiei) [inline ts/src/pro/bitmex.ts:607]: @pcriadoperez it crashed when I was testing it out

```
side': 'long',
  'stopLossPrice': None,
  'symbol': 'BTC/USDT:USDT',
  'takeProfitPrice': None,
  'timestamp': 1702022400208,
  'unrealizedPnl': '0'}]
Traceback (most recent call last):
  File "/Users/cjg/Git/ccxt10/ccxt/examples/py/cli.py", line 241, in <module>
    asyncio.run(main())
  File "/opt/homebrew/Cellar/python@3.11/3.11.5/Frameworks/Python.framework/Versions/3.11/lib/python3.11/asyncio/runners.py", line 190, in run
    return runner.run(main)
           ^^^^^^^^^^^^^^^^
  File "/opt/homebrew/Cellar/python@3.11/3.11.5/Frameworks/Python.framework/Versions/3.11/lib/python3.11/asyncio/runners.py", line 118, in run
    return self._loop.run_until_complete(task)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/homebrew/Cellar/python@3.11/3.11.5/Frameworks/Python.framework/Versions/3.11/lib/python3.11/asyncio/base_events.py", line 653, in run_until_complete
    return future.result()
           ^^^^^^^^^^^^^^^
  File "/Users/cjg/Git/ccxt10/ccxt/examples/py/cli.py", line 220, in main
    result = await result
             ^^^^^^^^^^^^
  File "/Users/cjg/Git/ccxt10/ccxt/python/ccxt/pro/bitmex.py", line 610, in watch_positions
    newPositions = await self.watch(url, messageHash, request, subscriptionHash)
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/cjg/Git/ccxt10/ccxt/python/ccxt/async_support/base/ws/fast_client.py", line 27, in handler
    self.handle_message(message)
  File "/Users/cjg/Git/ccxt10/ccxt/python/ccxt/async_support/base/ws/aiohttp_client.py", line 34, in handle_message
    self.handle_text_or_binary_message(message.data)
  File "/Users/cjg/Git/ccxt10/ccxt/python/ccxt/async_support/base/ws/aiohttp_client.py", line 29, in handle_text_or_binary_message
    self.on_message_callback(self, decoded)
  File "/Users/cjg/Git/ccxt10/ccxt/python/ccxt/pro/bitmex.py", line 1529, in handle_message
    return method(client, message)
           ^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/cjg/Git/ccxt10/ccxt/python/ccxt/pro/bitmex.py", line 771, in handle_positions
    position = self.parse_position(rawPosition)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/cjg/Git/ccxt10/ccxt/python/ccxt/async_support/bitmex.py", line 2121, in parse_position
    notionalString = Precise.string_abs(self.safe_string(position, 'foreignNotional', 'homeNotional'))
                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/cjg/Git/ccxt10/ccxt/python/ccxt/base/precise.py", line 227, in string_abs
    return str(Precise(string).abs())
               ^^^^^^^^^^^^^^^
  File "/Users/cjg/Git/ccxt10/ccxt/python/ccxt/base/precise.py", line 21, in __init__
    modifier = int(modifier)
               ^^^^^^^^^^^^^
```

Comment (pcriadoperez) [reply ts/src/pro/bitmex.ts:607]: hmm strange I stried in python both BTC/USD:BTC and BTC/USDT:USDT and both worked for me:
```py
[{'collateral': None,
  'contractSize': 1e-06,
  'contracts': 1000.0,
  'datetime': '2023-12-11T20:23:04.740Z',
  'entryPrice': None,
  'hedged': None,
  'id': '412475',
  'info': {'account': 412475,
           'currency': 'USDt',
           'currentQty': 1000,
           'foreignNotional': -40.91912,
           'liquidationPrice': 0.0,
           'maintMargin': 546650,
           'markPrice': 40919.12,
           'markValue': 40919120,
           'riskValue': 40919120,
           'symbol': 'XBTUSDT',
           'timestamp': '2023-12-11T20:23:04.740Z',
           'unrealisedPnl': 107620,
           'unrealisedRoePcnt': 0.2637},
  'initialMargin': None,
  'initialMarginPercentage': None,
  'lastPrice': None,
  'lastUpdateTimestamp': None,
  'leverage': None,
  'liquidationPrice': 0.0,
  'maintenanceMargin': '0.54665',
  'maintenanceMarginPercentage': None,
  'marginMode': 'isolated',
  'marginRatio': None,
  'markPrice': 40919.12,
  'notional': 40.91912,
  'percentage': None,
  'side': None,
  'stopLossPrice': None,
  'symbol': 'BTC/USDT:USDT',
  'takeProfitPrice': None,
  'timestamp': 1702326184740,
  'unrealizedPnl': '0.10762'}]
  ```

Comment (pcriadoperez) [reply ts/src/pro/bitmex.ts:607]: I wasn't able to reproduce but from the error log you posted I think I can see where it happens. I'll post a fix.

Comment (carlosmiei) [reply ts/src/pro/bitmex.ts:607]: @pcriadoperez No, I was using the sandbox as well

Comment (carlosmiei) [reply ts/src/pro/bitmex.ts:607]: @pcriadoperez this is easy to reproduce:
- start watching Positions ( p bitmex watchPositions --sandbox)
- change the position value (bitmex createOrder "SOL/USDT:USDT" market buy 1000  --sandbox)

Comment (carlosmiei) [reply ts/src/pro/bitmex.ts:607]: hum now that I'm debugging it, it does not crash...

```
➜  ccxt git:(bitmex/watchPositions) ✗  cd /Users/cjg/Git/ccxt10/ccxt ; /usr/bin/env /opt/homebrew/bin/python3 /Users/cjg/.vscode/extensions/ms-python.p
ython-2023.22.0/pythonFiles/lib/python/debugpy/adapter/../../debugpy/launcher 63668 -- /Users/cjg/Git/ccxt10/ccxt/examples/py/bybit-updated.py 
CCXT Version: 4.1.84
notional::-7.0771
notional::0.0
notional::0.0
notional::-7.0789
notional::-7.0776
notional::-14.1552
notional::-14.152
notional::-7.076
notional::-7.0759
notional::-7.0758
notional::-7.0752
notional::-7.0751
notional::-14.1502
notional::-14.1468
notional::-14.14
notional::-7.07
notional::-7.0689
```

Comment (carlosmiei) [reply ts/src/pro/bitmex.ts:607]: @pcriadoperez I found a different issue, after some time running the connection closes with a 1006 error, are we missing a ping or smth?

Comment (carlosmiei) [reply ts/src/pro/bitmex.ts:607]: for some reason isn't happening anymore..

**Comment by carlosmiei** [comment]
DEMO

```
p bitmex watchPositions --sandbox          
Python v3.11.5
CCXT v4.1.84
bitmex.watchPositions()
[{'collateral': None,
  'contractSize': 0.0001,
  'contracts': 0.0,
  'datetime': '2023-12-12T14:00:00.193Z',
  'entryPrice': None,
  'hedged': None,
  'id': '395724',
  'info': {'account': 395724,
           'commission': 0.00075,
           'crossMargin': True,
           'currency': 'USDt',
           'currentComm': 53078,
           'currentCost': 64000,
           'currentQty': 0,
           'foreignNotional': 0.0,
           'grossOpenPremium': 0,
           'homeNotional': 0.0,
           'initMargin': 0,
           'initMarginReq': 0.03,
           'isOpen': False,
           'leverage': 33.333333333333336,
           'liquidationPrice': None,
           'maintMargin': 0,
           'maintMarginReq': 0.015,
           'markPrice': None,
           'markValue': 0,
           'openOrderBuyCost': 0,
           'openOrderBuyPremium': 0,
           'openOrderBuyQty': 0,
           'openOrderSellCost': 0,
           'openOrderSellPremium': 0,
           'openOrderSellQty': 0,
           'openingQty': 0,
           'posComm': 0,
           'posCost': 0,
           'posCross': 0,
           'posLoss': 0,
           'posMaint': 0,
           'posMargin': 0,
           'prevRealisedPnl': -117078,
           'prevUnrealisedPnl': 0,
           'quoteCurrency': 'USDT',
           'realisedCost': 64000,
           'realisedPnl': -117078,
           'rebalancedPnl': 117078,
           'riskLimit': 1000000000000,
           'riskValue': 0,
           'symbol': 'SOLUSDT',
           'timestamp': '2023-12-12T14:00:00.193Z',
           'underlying': 'SOL',
           'unrealisedCost': 0,
           'unrealisedPnl': 0,
           'unrealisedPnlPcnt': 0.0,
           'unrealisedRoePcnt': 0.0},
  'initialMargin': 0.0,
  'initialMarginPercentage': 0.03,
  'lastPrice': None,
  'lastUpdateTimestamp': None,
  'leverage': 33.333333333333336,
  'liquidationPrice': None,
  'maintenanceMargin': '0',
  'maintenanceMarginPercentage': 0.015,
  'marginMode': 'cross',
  'marginRatio': None,
  'markPrice': None,
  'notional': 0.0,
  'percentage': 0.0,
  'side': 'long',
  'stopLossPrice': None,
  'symbol': 'SOL/USDT:USDT',
  'takeProfitPrice': None,
  'timestamp': 1702389600193,
  'unrealizedPnl': '0'},
 {'collateral': None,
  'contractSize': 100000000.0,
  'contracts': 0.0,
  'datetime': '2023-12-12T14:00:00.193Z',
  'entryPrice': None,
  'hedged': None,
  'id': '395724',
  'info': {'account': 395724,
           'commission': 0.00075,
           'crossMargin': True,
           'currency': 'XBt',
           'currentComm': 0,
           'currentCost': 0,
           'currentQty': 0,
           'foreignNotional': 0.0,
           'grossOpenPremium': 0,
           'homeNotional': 0.0,
           'initMargin': 0,
           'initMarginReq': 0.01,
           'isOpen': False,
           'leverage': 100.0,
           'liquidationPrice': None,
           'maintMargin': 0,
           'maintMarginReq': 0.0035,
           'markPrice': None,
           'markValue': 0,
           'openOrderBuyCost': 0,
           'openOrderBuyPremium': 0,
           'openOrderBuyQty': 0,
           'openOrderSellCost': 0,
           'openOrderSellPremium': 0,
           'openOrderSellQty': 0,
           'openingQty': 0,
           'posComm': 0,
           'posCost': 0,
           'posCross': 0,
           'posLoss': 0,
           'posMaint': 0,
           'posMargin': 0,
           'prevRealisedPnl': 0,
           'prevUnrealisedPnl': 0,
           'quoteCurrency': 'USD',
           'realisedCost': 0,
           'realisedPnl': 0,
           'rebalancedPnl': 0,
           'riskLimit': 20000000000,
           'riskValue': 0,
           'symbol': 'XBTUSD',
           'timestamp': '2023-12-12T14:00:00.193Z',
           'underlying': 'XBT',
           'unrealisedCost': 0,
           'unrealisedPnl': 0,
           'unrealisedPnlPcnt': 0.0,
           'unrealisedRoePcnt': 0.0},
  'initialMargin': 0.0,
  'initialMarginPercentage': 0.01,
  'lastPrice': None,
  'lastUpdateTimestamp': None,
  'leverage': 100.0,
  'liquidationPrice': None,
  'maintenanceMargin': '0',
  'maintenanceMarginPercentage': 0.0035,
  'marginMode': 'cross',
  'marginRatio': None,
  'markPrice': None,
  'notional': 0.0,
  'percentage': 0.0,
  'side': 'long',
  'stopLossPrice': None,
  'symbol': 'BTC/USD:BTC',
  'takeProfitPrice': None,
  'timestamp': 1702389600193,
  'unrealizedPnl': '0'},
 {'collateral': None,
  'contractSize': 1e-06,
  'contracts': 0.0,
  'datetime': '2023-12-12T14:00:00.193Z',
  'entryPrice': None,
  'hedged': None,
  'id': '395724',
  'info': {'account': 395724,
           'commission': 0.00075,
           'crossMargin': True,
           'currency': 'USDt',
           'currentComm': 0,
           'currentCost': 0,
           'currentQty': 0,
           'foreignNotional': 0.0,
           'grossOpenPremium': 0,
           'homeNotional': 0.0,
           'initMargin': 0,
           'initMarginReq': 0.01,
           'isOpen': False,
           'leverage': 100.0,
           'liquidationPrice': None,
           'maintMargin': 0,
           'maintMarginReq': 0.005,
           'markPrice': None,
           'markValue': 0,
           'openOrderBuyCost': 0,
           'openOrderBuyPremium': 0,
           'openOrderBuyQty': 0,
           'openOrderSellCost': 0,
           'openOrderSellPremium': 0,
           'openOrderSellQty': 0,
           'openingQty': 0,
           'posComm': 0,
           'posCost': 0,
           'posCross': 0,
           'posLoss': 0,
           'posMaint': 0,
           'posMargin': 0,
           'prevRealisedPnl': 10118730,
           'prevUnrealisedPnl': 0,
           'quoteCurrency': 'USDT',
           'realisedCost': 0,
           'realisedPnl': 0,
           'rebalancedPnl': 0,
           'riskLimit': 1000000000000,
           'riskValue': 0,
           'symbol': 'XBTUSDT',
           'timestamp': '2023-12-12T14:00:00.193Z',
           'underlying': 'XBT',
           'unrealisedCost': 0,
           'unrealisedPnl': 0,
           'unrealisedPnlPcnt': 0.0,
           'unrealisedRoePcnt': 0.0},
  'initialMargin': 0.0,
  'initialMarginPercentage': 0.01,
  'lastPrice': None,
  'lastUpdateTimestamp': None,
  'leverage': 100.0,
  'liquidationPrice': None,
  'maintenanceMargin': '0',
  'maintenanceMarginPercentage': 0.005,
  'marginMode': 'cross',
  'marginRatio': None,
  'markPrice': None,
  'notional': 0.0,
  'percentage': 0.0,
  'side': 'long',
  'stopLossPrice': None,
  'symbol': 'BTC/USDT:USDT',
  'takeProfitPrice': None,
  'timestamp': 1702389600193,
  'unrealizedPnl': '0'}]
```

File: ts/src/pro/bitmex.ts (hunk 1 above)
Comment (pcriadoperez) [reply ts/src/pro/bitmex.ts:607]: @carlosmiei , yes tried testing it from my side and stay subscribed for 30 minutes and didn't hit any 1006 error

**Pull request merged**